### DISCUSSION TITLE
Multiple Access Specifier Blocks

### DISCUSSION THREAD

**Discussion by xCyborg** [dotnet/csharplang#8882] [Language Ideas] [answered]
I always wanted C++ style Multiple Access Specifier Blocks for C#, it's probably one of the biggest features I miss in C# and I'm really disappointed it's not already implemented.

This proposal is to introduce **Multiple Access Specifier Blocks** in C# to allow grouping class or struct members under a single access modifier, similar to the existing feature in C++. This change aims to improve code organization, readability, and maintainability, especially for classes with multiple members sharing the same access level.

## Motivation
Currently, in C#, each class or struct member must explicitly declare its access modifier (e.g., `public`, `private`, `protected`, etc.). While this approach ensures clarity, it can become repetitive and verbose for classes with multiple members of the same access level. For example:

```csharp
public class Example
{
    public int PublicField1;
    public int PublicField2;
    public void PublicMethod1() {}
    public void PublicMethod2() {}

    private int PrivateField1;
    private int PrivateField2;
    private void PrivateMethod1() {}
    private void PrivateMethod2() {}
}
```

The repetition of `public` and `private` keywords for each member can make the code harder to read and maintain.

By introducing Multiple Access Specifier Blocks, developers can group members under a single access level, reducing repetition and improving clarity.

## Detailed Design
The proposed syntax mirrors that of C++, where access specifiers are defined as blocks with label-like modifiers:

```csharp
public class Example
{
    public:
        int PublicField1;
        int PublicField2;
        void PublicMethod1();
        void PublicMethod2();

    private:
        int PrivateField1;
        int PrivateField2;
        void PrivateMethod1();
        void PrivateMethod2();
}
```

### Proposed Syntax
- Introduce a colon `:` syntax after the access modifier (`public`, `private`, `protected`, etc.) to signify the start of a block.
- All members declared within the block inherit the specified access level until another block or explicit access modifier is encountered.

Example:

```csharp
public class Example
{
    public:
        int PublicField1;
        int PublicField2;
        void PublicMethod1();
        void PublicMethod2();

    private:
        int PrivateField1;
        int PrivateField2;
        void PrivateMethod1();
        void PrivateMethod2();
}
```

This is equivalent to the current syntax:

```csharp
public class Example
{
    public int PublicField1;
    public int PublicField2;
    public void PublicMethod1();
    public void PublicMethod2();

    private int PrivateField1;
    private int PrivateField2;
    private void PrivateMethod1();
    private void PrivateMethod2();
}
```

### Combined Modifiers
The feature could also support combined modifiers, such as `private protected`, `internal protected`, or `static public`. For example:

```csharp
public class Example
{
    private protected:
        int ProtectedInternalField;
        void ProtectedInternalMethod();

    static public:
        int StaticPublicField;
        void StaticPublicMethod();
}
```

This ensures flexibility while maintaining the benefits of grouped declarations.

### Rules and Constraints
1. **Block Termination:**
   - A block ends when a new access specifier is declared or when the class/struct body ends.
   - Explicitly declaring an access modifier for a member overrides the current block's access level.

2. **Nested Blocks:**
   - Nested types can have their own access specifier blocks.

3. **Backward Compatibility:**
   - Existing C# code will remain valid.
   - The feature is purely additive, meaning developers can continue to use explicit access modifiers for each member if they prefer.

4. **Reflection and Metadata:**
   - The underlying metadata and reflection output will remain unchanged. The compiler will treat block-declared members as if they were explicitly declared with the block's access modifier.

## Advantages
1. **Improved Readability:**
   - Reduces visual clutter by grouping members of the same access level.
   - Makes the structure of a class or struct more apparent.

2. **Reduced Repetition:**
   - Eliminates the need to repeatedly specify access modifiers for members sharing the same access level.

3. **Organizational Clarity:**
   - Encourages logical grouping of members based on their accessibility.

4. **Easier Maintenance:**
   - Changes to access levels are centralized and can be adjusted more easily.

## Implementation Considerations
1. **Compiler Updates:**
   - Update the C# compiler to parse and interpret access specifier blocks.
   - Ensure backward compatibility.

2. **IDE and Tooling Support:**
   - Update IntelliSense, syntax highlighting, and code formatting tools to recognize and support the new syntax.

3. **Error Handling:**
   - Provide clear error messages for misuse (e.g., missing colon, misplaced blocks).

## Conclusion
Introducing Multiple Access Specifier Blocks in C# would enhance the language's expressiveness and make code more concise and readable. While this feature draws inspiration from C++, it aligns well with C#'s focus on clarity and developer productivity and I suppose it wouldn't need any Runtime or IL modification (just simple syntactic sugar) and wouldn't also interfere with goto labels.

I hope we will finally consider this proposal.

**Comment by spydacarnage** [comment]
Sorry, I really don't like this at all.
If I see this in code:
```cs
void DoSomething() { }
```
I don't want to have to scroll up the class to find out whether it is public or private.  This can also lead to an accidental exposure of private members with a single `public:` declaration.

**Reply by xCyborg** [reply]
I didn't necessarily propose that it be used with functions or methods, for C++ functions it's mainly used with function signature declaration, which C# doesn't have, so it could be limited as an option for multiple consecutive member fields and properties instead where it makes perfect sense.
As for private members exposure it's primarily the developer's responsibility which should be mitigated by proper guidelines: if you use them, it's your responsibility to use them right.

**Reply by spydacarnage** [reply]
Your example explicitly showed method signatures, which is why I commented on them.

As for the private member exposure, `void DoSomething() { }` is already a valid `private` declaration, so it would be a gotcha for it to actually be `public` or something else due to a single line of code off-screen.

**Reply by CyrusNajmabadi** [reply]
> and properties 

Properties are also multi line, like methods.  (They really are just a pair of methods that allow a nicer calling convention at the call site).

**Comment by jubruckne** [comment]
I’d love this, although maybe with a slightly different syntax.

```
public class Example {
    public {
        int PublicField1;
        int PublicField2;
        void PublicMethod1();
        void PublicMethod2();
    }
    private {
        int PrivateField1;
        int PrivateField2;
        void PrivateMethod1();
        void PrivateMethod2();
     }

    public void ExplicitPublicMethod2();
    private int ExplicitPrivateField1;
}
```

> I don't` want to have to scroll up the class to find out whether it is public or private

IDEs will give you breadcrumbs so you’ll quickly see where you are. I find all the explicit access modifiers add noise when quickly scrolling through a file, scanning for a function I’m looking for.

**Reply by spydacarnage** [reply]
> IDEs will give you breadcrumbs so you’ll quickly see where you are. I find all the explicit access modifiers add noise when quickly scrolling through a file, scanning for a function I’m looking for.

Code isn't always only read in an IDE.  What about in a pull request diff?  Is this method that is changing public or private?  Or, worse still, what is this new `public {` block affecting?

**Comment by ufcpp** [comment, marked as answer]
https://github.com/dotnet/csharplang/discussions/1021
